llvm/lib/Support/Mustache.cpp

```diff
@@ -0,0 +1,489 @@
+                                 StringMap<Lambda> &Lambdas,
+                                 StringMap<SectionLambda> &SectionLambdas,
+                                 DenseMap<char, StringRef> &Escapes) {
+  LocalContext = Data;
+  Value Context = T == Root ? Data : findContext();
+  SmallString<128> Result;
+  switch (T) {
+  case Root: {
+    for (ASTNode *Child : Children)
+      Result += Child->render(Context, Allocator, Partials, Lambdas,
+                              SectionLambdas, Escapes);
+    return Result;
+  }
+  case Text:
+    return Body;
+  case Partial: {
+    if (Partials.find(Accessor[0]) != Partials.end()) {
+      ASTNode *Partial = Partials[Accessor[0]];
+      Result += Partial->render(Data, Allocator, Partials, Lambdas,
+                                SectionLambdas, Escapes);
+      addIndentation(Result, Indentation);
+    }
+    return Result;
+  }
+  case Variable: {
+    if (Lambdas.find(Accessor[0]) != Lambdas.end()) {
+      Lambda &L = Lambdas[Accessor[0]];
+      Value LambdaResult = L();
+      StringRef LambdaStr = printJson(LambdaResult);
+      Parser P = Parser(LambdaStr, Allocator);
+      ASTNode *LambdaNode = P.parse();
+      SmallString<128> RenderStr = LambdaNode->render(
+          Data, Allocator, Partials, Lambdas, SectionLambdas, Escapes);
+      return escapeString(RenderStr, Escapes);
+    }
+    return escapeString(printJson(Context), Escapes);
+  }
+  case UnescapeVariable: {
+    if (Lambdas.find(Accessor[0]) != Lambdas.end()) {
+      Lambda &L = Lambdas[Accessor[0]];
+      Value LambdaResult = L();
+      StringRef LambdaStr = printJson(LambdaResult);
+      Parser P = Parser(LambdaStr, Allocator);
+      ASTNode *LambdaNode = P.parse();
+      DenseMap<char, StringRef> EmptyEscapes;
+      return LambdaNode->render(Data, Allocator, Partials, Lambdas,
+                                SectionLambdas, EmptyEscapes);
+    }
+    return printJson(Context);
+  }
+  case Section: {
+    // Sections are not rendered if the context is falsey
+    bool IsLambda = SectionLambdas.find(Accessor[0]) != SectionLambdas.end();
+    if (isFalsey(Context) && !IsLambda)
+      return Result;
+    if (IsLambda) {
+      SectionLambda &Lambda = SectionLambdas[Accessor[0]];
+      Value Return = Lambda(RawBody);
+      if (isFalsey(Return))
+        return Result;
+      StringRef LambdaStr = printJson(Return);
+      Parser P = Parser(LambdaStr, Allocator);
+      ASTNode *LambdaNode = P.parse();
+      Result = LambdaNode->render(Data, Allocator, Partials, Lambdas,
+                                  SectionLambdas, Escapes);
+      return Result;
+    }
+    if (Context.getAsArray()) {
+      json::Array *Arr = Context.getAsArray();
+      for (Value &V : *Arr) {
+        for (ASTNode *Child : Children)
+          Result += Child->render(V, Allocator, Partials, Lambdas,
+                                  SectionLambdas, Escapes);
+      }
```

Comment:
This seems pretty expensive in terms of number of allocations. Especially since each SmallString is a minimum of 128 bytes based on the return type of `render()`. For now its probably OK, but we should revisit this, since its O(N**2).